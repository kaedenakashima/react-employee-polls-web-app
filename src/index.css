body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
}

.nav-header {
  display: flex;
  justify-content: space-between;
}

.nav-ttl {
  font-size: 16px;
}

.nav-items {
  display: flex;
}

.nav-item {
  display: flex;
  align-items: center;
  margin: 0 0.5em;
  text-decoration: none;
  color: black;
}

.nav-display {
  display: flex;
  align-items: center;
  padding: 4px 0 4px 8px;
  border-radius: 4px;
  background: #ededed;
  text-decoration: none;
  font-size: 12px;
  color: black;
  height: 20px;
  margin: 0.5em;
  font-weight: bold;
}

.nav-item:hover,
.nav-icon:hover {
  transform: scale(1.1);
}

.vote-result:hover,
.table-selected {
  cursor: pointer;
  text-decoration: underline;
}

.questions-table {
  font-family: Arial, Helvetica, sans-serif;
  border-collapse: collapse;
  width: fit-content;
  font-size: 12px;
  margin: 0 10px;
}

.ttl-table {
  width: 250px;
}

.questions-table td,
.questions-table th {
  padding: 8px;
  vertical-align: middle;
}

.questions-table tr {
  border: 1px solid black;
}

.questions-table tr:nth-child(even) {
  background-color: #f2f2f2;
}

.questions-table th {
  padding-top: 12px;
  padding-bottom: 12px;
  text-align: left;
  border: 1px solid black;
}

.cell-btn {
  margin: 0.7em;
}

.cell-btns {
  display: flex;
  margin: 0.7em;
}

.cell-vote-btn {
  border-radius: 2px;
  border: 1px solid black;
  font-size: 12px;
  height: 20px;
}

.user-icon {
  width: 38px;
  height: 38px;
  border-radius: 50%;
}

.user-icon-detail {
  width: 17px;
  height: 20px;
  border-radius: 50%;
  vertical-align: bottom;
}

.user-name {
  text-align: center;
  margin: auto 4px;
}

.progress-bar {
  max-width: 150px;
}

.progress-bar-container div {
  display: grid;
  grid-template-columns: 250px auto;
}

.progress-bar {
  display: flex;
}

.progress-bar span {
  font-size: 12px;
}

.progress-bar-optionOne,
.progress-bar-optionTwo {
  background: #343438;
  height: 15px;
}