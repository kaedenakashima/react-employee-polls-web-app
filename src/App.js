import React, {useState} from 'react'
import './index.css'
import {Routes, Route, Link, useNavigate, generatePath, Navigate} from 'react-router-dom'
import {generateId, questions, _getUsers, _getQuestions, _saveQuestion, _saveQuestionAnswer} from './_DATA'
import Leaderboard from './components/Leaderboard'
import Result from './components/Result'
import Login from './components/Login'
import NotFoundPage from './components/NotFoundPage'
import {useSelector, useDispatch } from 'react-redux'
import {configureStore, createSlice} from '@reduxjs/toolkit'
import TurnRightIcon from '@mui/icons-material/TurnRight'
import LogoutIcon from '@mui/icons-material/Logout'

var loginList = []
export const loginSlice = createSlice({
  name: 'loginedusers',
  initialState: loginList,
  reducers: {
    loginU: (state, action) => {
      state.push(action.payload)
    },
    addLoginUQ: (state, action) => {
      const {id, QId} = action.payload
      const y = state.find(x => x.id === id)
      if(y){
        y.questions = y.questions.concat(QId)
      }
    },
    addLoginUA: (state, action) => {
      const {id, aId, vote} = action.payload
      const y = state.find(x => x.id === id)
        y.aId.push(aId)
        y.answers.push(vote)
    },
    logoutU: (state, action) => {
      const {id} = action.payload
      const y = state.find(x => x.id === id)
      if(y){
        return state.filter(x=>x.id !== id)
      }
    }
  }
})
export const {loginU, addLoginUQ, addLoginUA, logoutU} = loginSlice.actions
export const loginsReducer = loginSlice.reducer
export const store = configureStore({
  reducer: {
    loginedusers:loginsReducer,
  }
})
export function Create() {
  const [txt1, setTxt1]=useState('')
  const [txt2, setTxt2]=useState('')
  const [error, setError]=useState('')
  _getQuestions()
  const loginedusers = useSelector((state)=> state.loginedusers)
  const dispatch = useDispatch()
  const navigate = useNavigate()
  const onClick2Add = async(e)=> {
    e.preventDefault()
    if (txt1 && txt2) {
      const generatedID = generateId()
      _saveQuestion({id: generatedID, author:loginedusers[0].id, optionOneText:txt1, optionTwoText:txt2})
      await dispatch(addLoginUQ({id: loginedusers[0].id, QId:generatedID}))
      setTxt1('')
      setTxt2('')
      _getQuestions()
      return setTimeout(() => {
        navigate(generatePath('/'))
        }, 3)
        
    } else {
      setError('*Please add both Choice1 and Choice2')
    }
  }
  return (
    <div style={{display:'inline-flex'}}> 
    <div>
      Do you rather choose <input data-testid='choice1' type='text' name='txt1' placeholder='Choice1' value={txt1} onChange={e => setTxt1(e.target.value)}/> or <input data-testid='choice2' type='text' name='txt2' placeholder='Choice2' value={txt2} onChange={e => setTxt2(e.target.value)}/>?
      <div data-testid='error' style={{color: 'red', fontSize:'12px'}}>{error}</div>
    </div>
    <div>
      <button
      onClick={onClick2Add}
      style={{background: 'none', border: 'none'}}>+</button>
    </div>
  </div>
  )
}
export function QuestionList() {
  var loginedusers = useSelector((state)=> state.loginedusers)
  _getQuestions()
  _getUsers()
  const dispatch = useDispatch()
  const [selectedTable, setSelectedTable] = useState('UnansweredQuestions')
  const [data2VoteUpdate, setData2VoteUpdate] = useState({})
  const [dataVoteUpdateId, setDataVoteUpdateId] = useState(null)
  const [vote, setVote]=useState()
  const navigate = useNavigate()
  const onClick2VoteUpdate = (e, id) => {
    e.preventDefault()
    setDataVoteUpdateId(id)
    const selectedQuestion = Object.values(questions).filter(x => x.id === id)
    const {vote} = selectedQuestion[0]
    setVote(vote)
  }
  const onClick2SaveVoteUpdate = async(e, id)=> {
    e.preventDefault()
    if(vote === Object.values(questions).find(x => x.id === id).optionOne.text){
      _saveQuestionAnswer ({ authedUser:loginedusers[0].id, qid:id, answer:'optionOne'})
      await dispatch(addLoginUA({id: loginedusers[0].id, aId:id, vote: 'optionOne'}))
    }
    else{
      _saveQuestionAnswer ({ authedUser:loginedusers[0].id, qid:id, answer:'optionTwo'})
      await dispatch(addLoginUA({id: loginedusers[0].id, aId:id, vote: 'optionTwo'}))
    }
    setData2VoteUpdate({id:id, vote:vote})
    _getQuestions()
    return setTimeout(() => {
    navigate(generatePath('/questions/:id', { id }))
    }, 3)
  }
  const onClick2CancelVoteUpdate = () => {
    setDataVoteUpdateId(null)
  }
  const isQMatched = (x) => {
    const y = loginedusers[0].aId.find(z => z === x.id)
    if(y){
      return true
    }
  }
  const onClick2QuestionResult = (e, id) => {
    e.preventDefault()
    return navigate(generatePath('/questions/:id', { id }))
  }
  const optionTabs = [
    {
      ttl: 'UnansweredQuestions',
      table: <table className='questions-table'>
      <tbody>
      <tr>
        <th></th>
        <th colSpan='2'>Question</th>
        <th>Created By</th>
        <th>Your Answer</th>
        <th>Result</th>
      </tr>
      {Object.values(questions).reverse().map((x, i)=> (
        <>
        {isQMatched(x)?
        undefined:<tr key={i} className='questions-row'>
          <td data-testid='questions'>{i+1}</td>
            <td className='vote-result' onClick={(e) => onClick2QuestionResult(e, x.id)}>Would you rather choose {x.optionOne.text} or {x.optionTwo.text}?</td>
            <td style={{display:'flex'}}>
            </td>
            <td>{new Date(x.timestamp).getUTCMonth() + 1}.{new Date(x.timestamp).getUTCDate()}.{new Date(x.timestamp).getUTCFullYear()} {x.author}</td>
              {dataVoteUpdateId === x.id ? (
              <td style={{display:'contents'}}>
              <form>
                <div>
                  <input data-testid='voteThisChoice' type='radio' id='voteChoice1' name='vote' value={x.optionOne.text} onChange={e => setVote(e.target.value)}/>
                  <label htmlFor='voteChoice1'>{x.optionOne.text}</label>
                  <input type='radio' id='voteChoice2' name='vote' value={x.optionTwo.text} onChange={e => setVote(e.target.value)}/>
                  <label htmlFor='voteChoice2'>{x.optionTwo.text}</label>
                </div>
              </form>
              <div className='cell-btns'>
                <button type='button' onClick={onClick2CancelVoteUpdate} style={{background: 'none', border: 'none', alignItems: 'center', color:'black',transform: 'rotate(180deg)'}}><TurnRightIcon/>
                </button><button className='cell-vote-btn' onClick={(e) => onClick2SaveVoteUpdate(e, x.id)} disabled={!vote}>Vote</button>
              </div>
            </td>
            ):(
              <td data-testid='answers' style={{textAlign:'center'}}>
              {isQMatched(x)? <span style={{textAlign:'center'}}>
              {loginedusers[0].answers[loginedusers[0].aId.findIndex((y) => y === x.id)]==='optionOne'? <>{x.optionOne.text}</>:<>{x.optionTwo.text}</>}
              </span>:<button className='cell-vote-btn' onClick={(e)=>onClick2VoteUpdate(e, x.id)}>Vote now!</button>}</td>
            )}
          <td>{x.optionOne.text}:{x.optionOne.votes.length}<br/>{x.optionTwo.text}:{x.optionTwo.votes.length}</td>
          </tr>
        }
        </>
      ))}
      </tbody>
  </table>
    },
    {
      ttl: 'AnsweredQuestions',
      table:<table className='questions-table'>
      <tbody>
      <tr>
        <th></th>
        <th colSpan='2'>Question</th>
        <th>Created By</th>
        <th>Your Answer</th>
        <th>Result</th>
      </tr>
      {Object.values(questions).reverse().map((x, i)=> (
        <>
        {isQMatched(x)?
        <tr key={i} className='questions-row'>
          <td data-testid='questions'>{i+1}</td>
            <td className='vote-result' onClick={(e) => onClick2QuestionResult(e, x.id)}>Would you rather choose {x.optionOne.text} or {x.optionTwo.text}?</td>
            <td style={{display:'flex'}}>
            </td>
            <td>{new Date(x.timestamp).getUTCMonth() + 1}.{new Date(x.timestamp).getUTCDate()}.{new Date(x.timestamp).getUTCFullYear()} {x.author}</td>
              {dataVoteUpdateId === x.id ? (
              <td style={{display:'tables'}}>
              <form>
                <div>
                  <input data-testid='voteThisChoice' type='radio' id='voteChoice1' name='vote' value={x.optionOne.text} onChange={e => setVote(e.target.value)}/>
                  <label htmlFor='voteChoice1'>{x.optionOne.text}</label>
                  <input type='radio' id='voteChoice2' name='vote' value={x.optionTwo.text} onChange={e => setVote(e.target.value)}/>
                  <label htmlFor='voteChoice2'>{x.optionTwo.text}</label>
                </div>
              </form>
              <div className='cell-btns'>
                <button type='button' onClick={onClick2CancelVoteUpdate} style={{background: 'none', border: 'none', alignItems: 'center', color:'black',transform: 'rotate(180deg)'}}><TurnRightIcon/>
                </button><button className='cell-vote-btn' onClick={(e) => onClick2SaveVoteUpdate(e, x.id)} disabled={!vote}>Vote</button>
              </div>
            </td>
            ):(
              <td data-testid='answers' style={{textAlign:'center'}}>
              {isQMatched(x)? <span style={{textAlign:'center'}}>
              {loginedusers[0].answers[loginedusers[0].aId.findIndex((y) => y === x.id)]==='optionOne'? <>{x.optionOne.text}</>:<>{x.optionTwo.text}</>}
              </span>:<button className='cell-vote-btn' onClick={(e)=>onClick2VoteUpdate(e, x.id)}>Vote now!</button>}</td>
            )}
          <td>{x.optionOne.text}:{x.optionOne.votes.length}<br/>{x.optionTwo.text}:{x.optionTwo.votes.length}</td>
          </tr>:undefined
        }
        </>
      ))}
      </tbody>
  </table>
    }
  ]
  return (
  <>
   <div>
        {optionTabs.map((x, i) => (
          <button
            key={i}
            onClick={() => setSelectedTable(x.ttl)}
            className={selectedTable === x.ttl ? 'table-selected':undefined}
            style={{display:'inline-block', margin: '0 4px', cursor:'pointer', fontSize:'12px', border:'none',background:'#fff'}}
          >
            {x.ttl}
          </button>
        ))}
      </div>
      <div>
        {optionTabs.map((x, i) => x.ttl === selectedTable && <div key={i}>{x.table}</div>)}
      </div>
  </>
  )
}
export const App = () =>  {
  const loginedusers = useSelector((state)=> state.loginedusers)
  const dispatch = useDispatch()
  const onClick3Logout = async (id) => {
    await dispatch(logoutU({id:id}))
  }
  return(
  <>
      {loginedusers.length === 0 ?(<>
      <HeaderNav2/>
      <Routes>
      <Route path='/' element={<Login/>}/>
      <Route path='*' element={<Login/>}/>
      </Routes>
      </>):undefined}
      {loginedusers.map((x, i)=> (<div key={i}>
       <HeaderNav onClick2Logout={() => onClick3Logout(x.id)}/>
      <Routes>
      <Route exact path='/' element={<QuestionList/>}/>
      <Route path='/questions/:id' element={<Result/>}/>
      <Route path='/add' element={<Create/>}/>
      <Route path='/leaderboard' element={<Leaderboard/>}/>
      <Route path="/404" element={<NotFoundPage/>}/>
      <Route path="*" element={<Navigate replace to="/404"/>}/>
      </Routes>
      </div>))}
  </>
  )
}
export const HeaderNav = ({onClick2Logout = () => {}}) => {
  var loginedusers = useSelector((state)=> state.loginedusers)
  return (
    <div className='nav-header'>
    <div className='nav-items'>
    <h1 className='nav-ttl'><b>React Udacity App</b></h1>
      <Link to='/' className='nav-item'>Vote</Link>
      <Link to='/add' className='nav-item'>Create Question</Link>
      <Link to='/leaderboard' className='nav-item'>Ranking</Link>
    </div>
      <div data-testid='loginuser' className='nav-display'>username:{loginedusers[0].name}<button onClick={() => onClick2Logout()} style={{background: 'none', border: 'none'}} className='nav-icon'><LogoutIcon/></button></div>
    </div>
  )
}
export const HeaderNav2 = () => {
  return (
    <div className='nav-header'>
    <h1 className='nav-ttl'><b>React Udacity App</b></h1>
    </div>
  )
}
export default App
