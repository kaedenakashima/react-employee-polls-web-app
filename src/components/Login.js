import React, {useState} from 'react'
import {users, _getUsers} from '../_DATA'
import {useDispatch} from 'react-redux'
import {loginU} from '../App'
export function Login() {
    _getUsers()
    const dispatch = useDispatch()
    const [txt, setTxt]=useState(null)
    const onClick2Login = async (e) => {
      e.preventDefault()
      if(txt){
        const selectedusers = Object.values(users).find(x => x.id === txt)
        await dispatch(loginU({id: selectedusers.id, name:selectedusers.name, aId:Object.keys(selectedusers.answers),answers:Object.values(selectedusers.answers), questions: selectedusers.questions}))
      }
    }
    return (
      <>
      <b>Login</b><br/>
      Please select your account<br/>
      User: <select data-testid='select' style={{width: '180px'}} onChange={e => setTxt(e.target.value)}>
      <option value=''>-----Please select--------</option>
        {Object.values(users).map((x, i)=> (<option data-testid='users' key={i} value={x.id}>{x.name}</option>))}
      </select><br/>
      <button data-testid="login" onClick={(e) => onClick2Login(e)} disabled={!txt}>Login</button>
      </>
    )
  }
export default Login
