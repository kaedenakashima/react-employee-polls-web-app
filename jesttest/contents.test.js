import React from 'react'
import '@testing-library/jest-dom'
import { render, screen, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import renderer from 'react-test-renderer'
import {Provider} from 'react-redux'
import {App, store, loginU, HeaderNav, HeaderNav2, QuestionList, Create} from '../src/App'
import Login from '../src/components/Login'
import {BrowserRouter} from 'react-router-dom'
import {_getUsers, _getQuestions, _saveQuestion, _saveQuestionAnswer} from '../src/_DATA'
test('test#1 Renders app title', () => {
    render(<BrowserRouter><Provider store={store}><HeaderNav2/></Provider></BrowserRouter>)
    const res = screen.getByText(/React Udacity App/i)
    expect(res).toBeInTheDocument()
})
test('test#2 App component snapshot test ', ()=> {
    const component = renderer.create(<BrowserRouter><Provider store={store}><App /></Provider></BrowserRouter>)
    let res = component.toJSON()
    expect(res).toMatchSnapshot()
})
const mockLoginUser = jest.fn()
test('test#3 login with correct user', async() => {
        let {getAllByTestId,getByTestId} = render(
            <BrowserRouter><Provider store={store}><Login setTxt={mockLoginUser} /></Provider></BrowserRouter>
        )
        const res1 = getAllByTestId('users')
        const res2 = await _getUsers()
        expect(Object.values(res1).length).toEqual(Object.values(res2).length)
        
        const res3 = getByTestId('select')
        fireEvent.change(res3, {target: {value:'tylermcginnis'}}) 
        expect(res3.value).toBe('tylermcginnis')
        const res4 = screen.getByRole('button', {name: 'Login'})
        fireEvent.click(res4)
})
const mockLoginedUser = [ { "id": "tylermcginnis", "name": "Tyler McGinnis", "aId": [ "vthrdm985a262al8qx3do", "xj352vofupe1dqz9emx13r" ], "answers": [ "optionOne", "optionTwo" ], "questions": [ "loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do" ] } ]
test('test#4 Header shows logined user name', async() => {
    const onClick2LoginSpy = async (e) => {
        e.preventDefault()
        if(txt){
            var selectedusers = Object.values(users).find(x => x.id === txt)
            await dispatch(loginU({id: selectedusers.id, name:selectedusers.name, aId:Object.keys(selectedusers.answers),answers:Object.values(selectedusers.answers), questions: selectedusers.question})) 
        }
    }
    const {getByTestId} = render(<BrowserRouter><Provider store={store}><Login setTxt={mockLoginUser} onClick2Login={onClick2LoginSpy} />
    <HeaderNav selecteduser={mockLoginedUser}/>
    </Provider></BrowserRouter>)
    const res1 = getByTestId('select')
    fireEvent.change(res1, {target: {value:'tylermcginnis'}}) 
    expect(res1.value).toBe('tylermcginnis')
    const res2 = screen.getByRole('button', {name: 'Login'})
    expect(res2).not.toBeDisabled()
    await userEvent.click(res2)
    expect(onClick2LoginSpy)
    expect(screen.getByText(/username:Tyler McGinnis/i)).toBeInTheDocument()
})
test('test#5 _getQuestions():Question displays all initial qestions.', async() => {
        const {getAllByTestId} = render(<BrowserRouter><Provider store={store}><QuestionList/></Provider></BrowserRouter>)
        const res1 = getAllByTestId('questions').length
        const x = screen.getByRole('button', {name: 'AnsweredQuestions'})
        fireEvent.click(x)
        const res2 = getAllByTestId('questions').length
        console.log('---------------------------------------',res2)
        const res3 = await _getQuestions()
        expect(res1+res2).toEqual(Object.values(res3).length)
        
})
test('test#6 _saveQuestion():Question is saved with correct format.', async() => {
    const mockChoice1 = jest.fn()
    const mockChoice2 = jest.fn()
    var {getAllByTestId, getByTestId} = render(<BrowserRouter><Provider store={store}><Create setTxt1={mockChoice1} setTxt2={mockChoice2}/></Provider></BrowserRouter>)
    const res1 = getByTestId('choice1')
    const res2 = getByTestId('choice2')
    fireEvent.change(res1, {target: {value:'Eat apple'}}) 
    fireEvent.change(res2, {target: {value:'Eat orange'}}) 
    expect(res1.value).toBe('Eat apple')
    expect(res2.value).toBe('Eat orange')
    const res3 =  await _saveQuestion({ optionOneText:res1.value, optionTwoText:res2.value, author:'tylermcginnis' })
    expect( Object.keys(res3)).toContain('id', 'timestamp', 'author', 'optionOne', 'optionTwo')
    const res4 = await _getQuestions()
    expect(Object.values(res4).length).toEqual(7)
    var {getAllByTestId, getByTestId} = render(<BrowserRouter><Provider store={store}><QuestionList /></Provider></BrowserRouter>)
    const res5 = getAllByTestId('questions').length
    const x = screen.getByRole('button', {name: 'AnsweredQuestions'})
    fireEvent.click(x)
    const res6 = getAllByTestId('questions').length
    expect(res5+res6).toEqual(Object.values(res4).length)
})
test('test#7 _saveQuestion(): error returns when user input wrong data', () => {
    const mockChoice1 = jest.fn()
    const mockChoice2 = jest.fn()
    const mockError = jest.fn()
    var {getByTestId} = render(<BrowserRouter><Provider store={store}><Create setTxt1={mockChoice1} setTxt2={mockChoice2} setError={mockError} /></Provider></BrowserRouter>)
    const a = getByTestId('choice1')
    const b = getByTestId('choice2')
    fireEvent.change(a, {target: {value:''}}) 
    fireEvent.change(b, {target: {value:'Eat orange'}}) 
    const c = screen.getByRole('button', {name: '+'})
    fireEvent.click(c)
    const d = getByTestId('error').textContent
    expect(d).toBe('*Please add both Choice1 and Choice2')
})
test('test#8 _getQuestions(): My answered item displays for each qestions correctly.', async() => {
    const questions = Object.values(await _getQuestions())
    render(<BrowserRouter><Provider store={store}><QuestionList/></Provider></BrowserRouter>)
    var loginedusers = [ { "id": "tylermcginnis", "name": "Tyler McGinnis", "aId": [ "vthrdm985a262al8qx3do", "xj352vofupe1dqz9emx13r" ], "answers": [ "optionOne", "optionTwo" ], "questions": [ "loxhs1bqm25b708cmbf3g", "vthrdm985a262al8qx3do" ] } ]
    const isQMatched = (a) => {
    const y = loginedusers[0].aId.find(x => x == a.id)
        if(y){
            return true
            }
        }
    var a = Object.values(questions)[4]
    var a2 = isQMatched(a)? loginedusers[0].answers[loginedusers[0].aId.findIndex((y) => y == a.id)]==='optionOne'? a.optionOne.text:a.optionTwo.text:'vote now'
    expect(a2).toBe('find $50 yourself')
    a = Object.values(questions)[5]
    a2 = isQMatched(a)? loginedusers[0].answers[loginedusers[0].aId.findIndex((y) => y == a.id)]==='optionOne'? a.optionOne.text:a.optionTwo.text:'vote now'
    expect(a2).toBe('write Swift')
})
test('test#9 _saveQuestionAnswer(): answer is saved with correct format', async () => {
    const mockVote = jest.fn()
    const mockId = jest.fn()
    var {getByTestId} = render(<BrowserRouter><Provider store={store}><QuestionList setData2VoteUpdate={mockVote} setData2VoteUpdateId={mockId} /></Provider></BrowserRouter>)
    fireEvent.click(screen.getAllByRole('button', {name: 'Vote now!'})[0])
    fireEvent.change(getByTestId('voteThisChoice'), {target: {value:'optionOne'}})
    fireEvent.click(screen.getByRole('button', {name: 'Vote'}))
    var answers = await _getUsers()
    var questions = await _getQuestions()
    expect(Object.keys(Object.values(answers)[1].answers).length).toEqual(2)
    expect(Object.values(questions)[0].optionOne.votes.length).toEqual(1)
    const res1 = await _saveQuestionAnswer({ authedUser:'tylermcginnis', qid: Object.values(questions)[0].id, answer:'optionOne'})
    expect(res1).toBe('ok!');
    answers = await _getUsers()
    questions = await _getQuestions()
    expect(Object.keys(Object.values(answers)[1].answers).length).toEqual(3)
    expect(Object.values(questions)[0].optionOne.votes.length).toEqual(2)
})
test('test#10 _saveQuestionAnswer(): button is disabled when user did not select a choice', () => {
    const mockVote = jest.fn()
    const mockId = jest.fn()
    var {getByTestId} = render(<BrowserRouter><Provider store={store}><QuestionList setData2VoteUpdate={mockVote} setData2VoteUpdateId={mockId} /></Provider></BrowserRouter>)
    fireEvent.click(screen.getAllByRole('button', {name: 'Vote now!'})[0])
    fireEvent.change(getByTestId('voteThisChoice'), {target: {value:''}})
    expect(screen.getByRole('button', {name: 'Vote'})).toHaveAttribute('disabled');
})