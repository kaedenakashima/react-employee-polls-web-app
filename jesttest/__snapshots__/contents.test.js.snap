// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test#2 App component snapshot test  1`] = `
Array [
  <div
    className="nav-header"
  >
    <h1
      className="nav-ttl"
    >
      <b>
        React Udacity App
      </b>
    </h1>
  </div>,
  <b>
    Login
  </b>,
  <br />,
  "Please select your account",
  <br />,
  "User: ",
  <select
    data-testid="select"
    onChange={[Function]}
    style={
      Object {
        "width": "180px",
      }
    }
  >
    <option
      value=""
    >
      -----Please select--------
    </option>
    <option
      data-testid="users"
      value="sarahedo"
    >
      Sarah Edo
    </option>
    <option
      data-testid="users"
      value="tylermcginnis"
    >
      Tyler McGinnis
    </option>
    <option
      data-testid="users"
      value="johndoe"
    >
      John Doe
    </option>
  </select>,
  <br />,
  <button
    data-testid="login"
    disabled={true}
    onClick={[Function]}
  >
    Login
  </button>,
]
`;
